{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "d476b993-8391-4e84-bf50-90b4e2badff5",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "'getOption(\"repos\")' replaces Bioconductor standard repositories, see\n",
      "'help(\"repositories\", package = \"BiocManager\")' for details.\n",
      "Replacement repositories:\n",
      "    CRAN: https://cran.r-project.org\n",
      "\n",
      "Bioconductor version 3.20 (BiocManager 1.30.27), R 4.4.3 (2025-02-28)\n",
      "\n",
      "Warning message:\n",
      "“package(s) not installed when version(s) same as or greater than current; use\n",
      "  `force = TRUE` to re-install: 'org.Hs.eg.db'”\n",
      "Old packages: 'base64enc', 'BH', 'BiocParallel', 'Biostrings',\n",
      "  'clusterProfiler', 'data.table', 'dbplyr', 'DOSE', 'dplyr', 'dtplyr',\n",
      "  'enrichplot', 'fgsea', 'GenomeInfoDb', 'ggplot2', 'httr', 'igraph',\n",
      "  'IRanges', 'lattice', 'lubridate', 'MatrixGenerics', 'RSQLite',\n",
      "  'SparseArray', 'viridisLite', 'vroom', 'yulab.utils'\n",
      "\n"
     ]
    }
   ],
   "source": [
    "if (!requireNamespace(\"BiocManager\", quietly = TRUE))\n",
    "    install.packages(\"BiocManager\")\n",
    "\n",
    "BiocManager::install(\"org.Hs.eg.db\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "01fb367d-6a8e-4850-80e4-16b7ab374153",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: S4Vectors\n",
      "\n",
      "Loading required package: stats4\n",
      "\n",
      "Loading required package: BiocGenerics\n",
      "\n",
      "\n",
      "Attaching package: ‘BiocGenerics’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    anyDuplicated, aperm, append, as.data.frame, basename, cbind,\n",
      "    colnames, dirname, do.call, duplicated, eval, evalq, Filter, Find,\n",
      "    get, grep, grepl, intersect, is.unsorted, lapply, Map, mapply,\n",
      "    match, mget, order, paste, pmax, pmax.int, pmin, pmin.int,\n",
      "    Position, rank, rbind, Reduce, rownames, sapply, saveRDS, setdiff,\n",
      "    table, tapply, union, unique, unsplit, which.max, which.min\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘S4Vectors’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:utils’:\n",
      "\n",
      "    findMatches\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    expand.grid, I, unname\n",
      "\n",
      "\n",
      "Loading required package: IRanges\n",
      "\n",
      "Loading required package: GenomicRanges\n",
      "\n",
      "Loading required package: GenomeInfoDb\n",
      "\n",
      "Loading required package: SummarizedExperiment\n",
      "\n",
      "Loading required package: MatrixGenerics\n",
      "\n",
      "Loading required package: matrixStats\n",
      "\n",
      "\n",
      "Attaching package: ‘MatrixGenerics’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,\n",
      "    colCounts, colCummaxs, colCummins, colCumprods, colCumsums,\n",
      "    colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,\n",
      "    colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,\n",
      "    colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,\n",
      "    colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,\n",
      "    colWeightedMeans, colWeightedMedians, colWeightedSds,\n",
      "    colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,\n",
      "    rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,\n",
      "    rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,\n",
      "    rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,\n",
      "    rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,\n",
      "    rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,\n",
      "    rowWeightedMads, rowWeightedMeans, rowWeightedMedians,\n",
      "    rowWeightedSds, rowWeightedVars\n",
      "\n",
      "\n",
      "Loading required package: Biobase\n",
      "\n",
      "Welcome to Bioconductor\n",
      "\n",
      "    Vignettes contain introductory material; view with\n",
      "    'browseVignettes()'. To cite Bioconductor, see\n",
      "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘Biobase’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:MatrixGenerics’:\n",
      "\n",
      "    rowMedians\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    anyMissing, rowMedians\n",
      "\n",
      "\n",
      "Loading required package: AnnotationDbi\n",
      "\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "library(DESeq2)\n",
    "library(tximport)\n",
    "library(org.Hs.eg.db)\n",
    "library(AnnotationDbi)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "14ce06f2-efa8-4d7d-9988-b64e1dbb61a7",
   "metadata": {},
   "source": [
    "# 1- Read data "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "6dff3d17-d1ee-4f73-8da5-b324e3927852",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Samples: 6\n",
      "Genes: 2379\n",
      "Conditions: WT vs KO\n"
     ]
    }
   ],
   "source": [
    "gene_counts <- readRDS(\"~/genomics//ch2_project/counts/tximport.rds\")\n",
    " \n",
    "sample_info <- read.table(\"~/genomics/ch2_project/counts/sample_info.tsv\", header = TRUE, sep = \"\\t\")\n",
    "\n",
    "# Set factor levels (WT as reference)\n",
    "sample_info$condition <- factor(sample_info$condition, levels = c(\"WT\", \"KO\"))\n",
    "\n",
    "cat(paste0(\"Samples: \", nrow(sample_info), \"\\n\"))\n",
    "cat(paste0(\"Genes: \", nrow(gene_counts$counts), \"\\n\"))\n",
    "cat(paste0(\"Conditions: \", paste(levels(sample_info$condition), collapse = \" vs \"), \"\\n\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0722e72a-2156-489a-99f0-f682d3eca14d",
   "metadata": {},
   "source": [
    "# 2- Create DESeq dataset "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "698db714-402c-45ba-972c-9318eba158ea",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "using counts and average transcript lengths from tximport\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Genes after filtering (>=10 counts in >=3 samples): 922\n"
     ]
    }
   ],
   "source": [
    "dds <- DESeqDataSetFromTximport(gene_counts, colData = sample_info, design = ~ condition)\n",
    "\n",
    "# Pre-filtering: Remove genes with very low counts (genes with less than 10 counts in 3 or more samples will be removed)\n",
    "keep <- rowSums(counts(dds) >= 10) >= 3 \n",
    "dds <- dds[keep, ]\n",
    "cat(paste0(\"Genes after filtering (>=10 counts in >=3 samples): \", nrow(dds), \"\\n\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "55638eb9-55a5-403c-aacd-81991cd58df5",
   "metadata": {},
   "source": [
    "# 3- Run DESeq "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "43ee6fc2-a645-4073-bc49-dbbaa2b27c8d",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "estimating size factors\n",
      "\n",
      "using 'avgTxLength' from assays(dds), correcting for library size\n",
      "\n",
      "estimating dispersions\n",
      "\n",
      "gene-wise dispersion estimates\n",
      "\n",
      "mean-dispersion relationship\n",
      "\n",
      "final dispersion estimates\n",
      "\n",
      "fitting model and testing\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "out of 922 with nonzero total read count\n",
      "adjusted p-value < 0.1\n",
      "LFC > 0 (up)       : 163, 18%\n",
      "LFC < 0 (down)     : 144, 16%\n",
      "outliers [1]       : 0, 0%\n",
      "low counts [2]     : 0, 0%\n",
      "(mean count < 9)\n",
      "[1] see 'cooksCutoff' argument of ?results\n",
      "[2] see 'independentFiltering' argument of ?results\n",
      "\n",
      "Total significant genes: 41\n",
      "Upregulated in KO: 32\n",
      "Downregulated in KO: 9\n"
     ]
    }
   ],
   "source": [
    "dds <- DESeq(dds)\n",
    "                res <- results(dds, contrast = c(\"condition\", \"KO\", \"WT\"))\n",
    "res <- res[order(res$padj), ]\n",
    "\n",
    "summary(res)\n",
    "\n",
    "# Significant: padj < 0.05 and |log2FC| > 1\n",
    "sig_genes <- subset(res, padj < 0.05 & abs(log2FoldChange) > 1)\n",
    "sig_genes <- sig_genes[order(sig_genes$padj), ]\n",
    "\n",
    "cat(paste0(\"Total significant genes: \", nrow(sig_genes), \"\\n\"))\n",
    "cat(paste0(\"Upregulated in KO: \", sum(sig_genes$log2FoldChange > 0, na.rm = TRUE), \"\\n\"))\n",
    "cat(paste0(\"Downregulated in KO: \", sum(sig_genes$log2FoldChange < 0, na.rm = TRUE), \"\\n\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "301caaf7-0c44-48c3-9536-870c1c4b46f2",
   "metadata": {},
   "source": [
    "# 4- Save data "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "c013d66f-77e2-4a78-bcce-65d566a4469c",
   "metadata": {},
   "outputs": [],
   "source": [
    "dir.create(\"~/genomics/ch2_project/results/tables\", showWarnings = FALSE, recursive = TRUE)\n",
    "\n",
    "# All results\n",
    "all_res <- as.data.frame(res)\n",
    "all_res$gene_id <- rownames(all_res)\n",
    "all_res <- all_res[, c(\"gene_id\", \"baseMean\", \"log2FoldChange\", \"lfcSE\", \"stat\", \"pvalue\", \"padj\")]\n",
    "write.csv(all_res, \"~/genomics/ch2_project/results/tables/deseq2_all_results.csv\", row.names = FALSE)\n",
    "\n",
    "# Significant genes only\n",
    "sig_res <- as.data.frame(sig_genes)\n",
    "sig_res$gene_id <- rownames(sig_res)\n",
    "sig_res <- sig_res[, c(\"gene_id\", \"baseMean\", \"log2FoldChange\", \"lfcSE\", \"stat\", \"pvalue\", \"padj\")]\n",
    "write.csv(sig_res, \"~/genomics//ch2_project/results/tables/deseq2_significant.csv\", row.names = FALSE)\n",
    "\n",
    "# Normalized counts\n",
    "norm_counts <- as.data.frame(counts(dds, normalized = TRUE))\n",
    "norm_counts$gene_id <- rownames(norm_counts)\n",
    "write.csv(norm_counts, \"~/genomics/ch2_project/results/tables/normalized_counts.csv\", row.names = FALSE)\n",
    "\n",
    "# Save DESeq2 object for later use\n",
    "saveRDS(dds, \"~/genomics/ch2_project/results/deseq2_object.rds\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "792b89ed-0ea3-44e8-b4ac-1fbb86524f90",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "'select()' returned 1:many mapping between keys and columns\n",
      "\n",
      "'select()' returned 1:1 mapping between keys and columns\n",
      "\n",
      "'select()' returned 1:many mapping between keys and columns\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Function to map Ensembl IDs → Gene Symbols\n",
    "convert_ensembl_to_symbol <- function(ids) {\n",
    "  clean_ids <- gsub(\"\\\\..*\", \"\", ids)  # remove version numbers if any \n",
    "  symbols <- mapIds(org.Hs.eg.db,\n",
    "                    keys = clean_ids,\n",
    "                    column = \"SYMBOL\",\n",
    "                    keytype = \"ENSEMBL\",\n",
    "                    multiVals = \"first\")\n",
    "  # Replace NA with original Ensembl ID\n",
    "  symbols[is.na(symbols)] <- clean_ids[is.na(symbols)]\n",
    "  return(symbols)\n",
    "}\n",
    "\n",
    "# Map gene symbols for results\n",
    "all_res$gene_symbol <- convert_ensembl_to_symbol(all_res$gene_id)\n",
    "sig_res$gene_symbol <- convert_ensembl_to_symbol(sig_res$gene_id)\n",
    "norm_counts$gene_symbol <- convert_ensembl_to_symbol(norm_counts$gene_id)\n",
    "\n",
    "# move gene_symbol column to first position\n",
    "all_res <- all_res[, c(\"gene_symbol\", setdiff(names(all_res), \"gene_symbol\"))]\n",
    "sig_res <- sig_res[, c(\"gene_symbol\", setdiff(names(sig_res), \"gene_symbol\"))]\n",
    "norm_counts <- norm_counts[, c(\"gene_symbol\", setdiff(names(norm_counts), \"gene_symbol\"))]\n",
    "\n",
    "all_res <- all_res[, setdiff(names(all_res), \"gene_id\")]\n",
    "sig_res <- sig_res[, setdiff(names(sig_res), \"gene_id\")]\n",
    "norm_counts <- norm_counts[, setdiff(names(norm_counts), \"gene_id\")]\n",
    "\n",
    "# Save tables\n",
    "write.csv(all_res, \"~/genomics/ch2_project/results/tables/deseq2_all_results_genesymbol.csv\", row.names = FALSE)\n",
    "write.csv(sig_res, \"~/genomics/ch2_project/results/tables/deseq2_significant_genesymbol.csv\", row.names = FALSE)\n",
    "write.csv(norm_counts, \"~/genomics/ch2_project/results/tables/normalized_counts_genesymbol.csv\", row.names = FALSE)\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.4.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
